src/api_server/src/request/mmds.rs

```diff
@@ -1,13 +1,53 @@
 // Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 // SPDX-License-Identifier: Apache-2.0
 
+use super::super::VmmAction;
+use micro_http::StatusCode;
 use request::{Body, Error, ParsedRequest};
+use vmm::vmm_config::mmds::MmdsConfig;
 
 pub fn parse_get_mmds() -> Result<ParsedRequest, Error> {
     Ok(ParsedRequest::GetMMDS)
 }
 
-pub fn parse_put_mmds(body: &Body) -> Result<ParsedRequest, Error> {
+pub fn parse_put_mmds(body: &Body, first_token: Option<&&str>) -> Result<ParsedRequest, Error> {
+    let mut config_request = false;
+    if let Some(config_path) = first_token {
+        if *config_path == "config" {
+            config_request = true;
+        } else {
+            return Err(Error::Generic(
+                StatusCode::BadRequest,
+                format!("Unrecognized PUT request path `{}`.", *config_path),
+            ));
+        }
+    }
+
+    if config_request {
+        let mmds_config =
+            serde_json::from_slice::<MmdsConfig>(body.raw()).map_err(Error::SerdeJson)?;
+
+        let mmds_ipv4_addr_pool = mmds_config.ipv4_addr_pool();
+        if mmds_ipv4_addr_pool.is_err() {
+            // Safe, checked above.
+            return Err(Error::MmdsConfig(mmds_ipv4_addr_pool.err().unwrap()));
```

Comment:
I don't think we should wrap the error returned by `ip4_addr_pool()` in yet another MMDS specific error. The error returned by the aforementioned function should give us the necessary MMDS information and we could just wrap it in the `BadRequest` error type like you already do in the `is_empty()` scenario.
Also it looks like `ipv4_addr_pool()` doesn't have a single and clear purpose. In the context of parsing the HTTP request you are using it just to see if the `Vec` of addresses is not empty and in the vmm builder context you are actually using those addresses to configure a net device.
Also, I'm not so sure, when parsing the request body, is it okay to throw a `BadRequest` error if this `Vec` is empty? When you are attaching a net device it looks like the `empty` array scenario is accepted.
Anyway, I believe that the `ipv4_addr_pool()` logic should be a little bit rethought.
Actually, how about changing the type of `ipv4_address_pool` from `Vec<String>` to `Vec<Ipv4Addr>` (`Ipv4Addr` is a "standard" type like `PathBuf`, so it should make sense). I imagine that `serde` will take care of the parsing logic and throwing the right errors if those strings are invalid in terms of ipv4 addresses. This way, you won't need the `ipv4_addr_pool()` anymore, you can just create a getter for the `ipv4_address_pool` field, that you can call when attaching a net device. Also the whole error handling will be greatly simplified this way (no need, for example, for the `IPv4ParseError` enum variant). The code would look something like this:

request/mmds.rs:

```rust
pub fn parse_put_mmds(body: &Body, request_type: Option<&&str>) -> Result<ParsedRequest, Error> {
    match request_type {
        Some(&config_path) => match config_path {
            "config" => {
                return Ok(ParsedRequest::Sync(VmmAction::SetMmdsConfiguration(
                    serde_json::from_slice::<MmdsConfig>(body.raw()).map_err(|e| {
                        Error::SerdeJson(e)
                    })?,
                )));
            }
            _ => {
                return Err(Error::Generic(
                    StatusCode::BadRequest,
                    format!("Unrecognized PUT request path `{}`.", config_path),
                ))
            }
        },
        None => {
            return Ok(ParsedRequest::PutMMDS(
                serde_json::from_slice(body.raw()).map_err(Error::SerdeJson)?,
            ))
        }
    }
}

```
vmm_config/mmds.rs:

```rust
pub struct MmdsConfig {
    /// Pool of MMDS endpoints.
    ipv4_address_pool: Vec<Ipv4Addr>,
}

impl MmdsConfig {
    pub fn ipv4_address_poll(&self) -> Vec<Ipv4Addr> {
        self.ipv4_address_pool.clone()
    }
}
```

 What do you think? :D